src/services/executionEnvironment.ts

```diff
@@ -0,0 +1,308 @@
+        this.context.workspaceFolder.uri
+      );
+      this._container_image = settings.executionEnvironment.image;
+      this._container_engine = settings.executionEnvironment.containerEngine;
+      if (this._container_engine === 'auto') {
+        for (const ce of ['podman', 'docker']) {
+          try {
+            child_process.execSync(`which ${ce}`, {
+              encoding: 'utf-8',
+            });
+            this._container_engine = <IContainerEngine>ce;
+            this.connection.console.log(`Container engine set to: '${ce}'`);
+            break;
+          } catch (error) {
+            this.connection.console.info(`Container engine '${ce}' not found`);
+          }
+        }
+      } else {
+        try {
+          child_process.execSync(`which ${this._container_engine}`, {
+            encoding: 'utf-8',
+          });
+        } catch (error) {
+          this.connection.window.showErrorMessage(
+            `Container engine '${this._container_engine}' not found. Failed with error '${error}'`
+          );
+          return;
+        }
+      }
+      if (!['podman', 'docker'].includes(this._container_engine)) {
+        this.connection.window.showInformationMessage(
+          'No valid container engine found.'
+        );
+        return;
+      }
+      const imagePuller = new ImagePuller(
+        this.connection,
+        this.context,
+        this._container_engine,
+        this._container_image,
+        settings.executionEnvironment.pullPolicy
+      );
+      const setupDone = await imagePuller.setupImage();
+      if (!setupDone) {
+        this.connection.console.error(
+          `Execution environment image ${this._container_image} setup failed.`
+        );
+        return;
+      }
+      this.fetchPluginDocs();
+    } catch (error) {
+      if (error instanceof Error) {
+        this.connection.window.showErrorMessage(error.message);
+      } else {
+        this.connection.console.error(
+          `Exception in ExecutionEnvironment service: ${JSON.stringify(error)}`
+        );
+      }
+    }
+  }
+
+  async fetchPluginDocs(): Promise<void> {
+    const ansibleConfig = await this.context.ansibleConfig;
+    const containerName = `${this._container_image.replace(
+      /[^a-z0-9]/gi,
+      '_'
+    )}`;
+
+    try {
+      const containerImageIdCommand = `${this._container_engine} images ${this._container_image} --format="{{.ID}}" | head -n 1`;
+      this.connection.console.log(containerImageIdCommand);
+      this._container_image_id = child_process
+        .execSync(containerImageIdCommand, {
+          encoding: 'utf-8',
+        })
+        .trim();
+      const hostCacheBasePath = path.resolve(
+        `${process.env.HOME}/.cache/ansible-language-server/${containerName}/${this._container_image_id}`
+      );
+
+      const isContainerRunning = this.runContainer(containerName);
+      if (!isContainerRunning) {
+        return;
```

Comment:
maybe show an error message?

Reply:
Will add error user feedback in `runContainer` as we know the exact error why it fails in that method

Reply:
Fixed